test/unit/codegen/codegen_utils.cpp

```diff
@@ -0,0 +1,218 @@
+                float_constant);
+            REQUIRE(nmodl_constant_result == codegen_output);
+        }
+    }
+
+    GIVEN("Double constants in scientific notation as strings") {
+        std::string double_constant_with_e = "1e+18";
+        std::string double_constant_with_e_minus = "1e-18";
+        std::string double_constant_with_decimal_point_e = ".123e18";
+        std::string double_constant_with_E = "1E18";
+
+
+        std::string result_double_constant_with_e = "1d+18";
+        std::string result_double_constant_with_e_minus = "1d-18";
+        std::string result_double_constant_with_decimal_point_e = ".123d18";
+        std::string result_double_constant_with_E = "1d18";
+
+        THEN("Codegen ISPC Visitor prints doubles with scientific notation") {
+            auto nmodl_constant_result = codegen::utils::format_double_string<CodegenIspcVisitor>(
+                double_constant_with_e);
+            REQUIRE(nmodl_constant_result == result_double_constant_with_e);
+            nmodl_constant_result = codegen::utils::format_double_string<CodegenIspcVisitor>(
+                double_constant_with_e_minus);
+            REQUIRE(nmodl_constant_result == result_double_constant_with_e_minus);
+            nmodl_constant_result = codegen::utils::format_double_string<CodegenIspcVisitor>(
+                double_constant_with_decimal_point_e);
+            REQUIRE(nmodl_constant_result == result_double_constant_with_decimal_point_e);
+            nmodl_constant_result = codegen::utils::format_double_string<CodegenIspcVisitor>(
+                double_constant_with_E);
+            REQUIRE(nmodl_constant_result == result_double_constant_with_E);
+        }
+    }
+
+    GIVEN("Float constants in scientific notation as strings") {
+        std::string float_constant_with_e = "1e+18";
+        std::string float_constant_with_e_minus = "1e-18";
+        std::string float_constant_with_decimal_point_e = ".123e18";
+        std::string float_constant_with_E = "1E18";
+
+        THEN("Codegen ISPC Visitor prints doubles with scientific notation") {
+            auto nmodl_constant_result = codegen::utils::format_float_string<CodegenIspcVisitor>(
+                float_constant_with_e);
+            REQUIRE(nmodl_constant_result == float_constant_with_e);
+            nmodl_constant_result = codegen::utils::format_float_string<CodegenIspcVisitor>(
+                float_constant_with_e_minus);
+            REQUIRE(nmodl_constant_result == float_constant_with_e_minus);
+            nmodl_constant_result = codegen::utils::format_float_string<CodegenIspcVisitor>(
+                float_constant_with_decimal_point_e);
+            REQUIRE(nmodl_constant_result == float_constant_with_decimal_point_e);
+            nmodl_constant_result = codegen::utils::format_float_string<CodegenIspcVisitor>(
+                float_constant_with_E);
+            REQUIRE(nmodl_constant_result == float_constant_with_E);
```

Comment:
Just a nitpick : as this pattern is getting repeated a lot in this file, one can think of wrapper function that accepts input vector (of numbers as strings) and output vector (of numbers as string) and then you iterate on each element of vector and compare the with an element of output vector. This will reduce the code in this test ?  (we use similar approach in [the parser tests](https://github.com/BlueBrain/nmodl/blob/1089bc1509e7ae8cc9a76e0d527c3f738014c47b/test/unit/parser/parser.cpp#L136)).

Reply:
I was thinking something similar in the beginning when I didn't have a lot of checks, so I thought it wasn't worth it. Then I started adding cases though, so it seems like a good idea now :stuck_out_tongue:

Reply:
Done. I hope you think it's better like this